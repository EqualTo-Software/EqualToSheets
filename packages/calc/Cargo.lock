# This file is automatically @generated by Cargo.
# It is not intended for manual editing.
version = 3

[[package]]
name = "adler"
version = "1.0.2"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "f26201604c87b1e01bd3d98f8d5d9a8fcbb815e8cedb41ffccbeb4bf593a35fe"

[[package]]
name = "aho-corasick"
version = "0.7.19"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "b4f55bd91a0978cbfd91c457a164bab8b4001c833b7f323132c0a4e1922dd44e"
dependencies = [
 "memchr",
]

[[package]]
name = "android_system_properties"
version = "0.1.5"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "819e7219dbd41043ac279b19830f2efc897156490d7fd6ea916720117ee66311"
dependencies = [
 "libc",
]

[[package]]
name = "autocfg"
version = "1.1.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "d468802bab17cbc0cc575e9b053f41e72aa36bfa6b7f55e3529ffa43161b97fa"

[[package]]
name = "bumpalo"
version = "3.11.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "c1ad822118d20d2c234f427000d5acc36eabe1e29a348c89b63dd60b13f28e5d"

[[package]]
name = "byteorder"
version = "1.4.3"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "14c189c53d098945499cdfa7ecc63567cf3886b3332b312a5b4585d8d3a6a610"

[[package]]
name = "cfg-if"
version = "1.0.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "baf1de4339761588bc0619e3cbc0120ee582ebb74b53b4efbf79117bd2da40fd"

[[package]]
name = "chrono"
version = "0.4.23"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "16b0a3d9ed01224b22057780a37bb8c5dbfe1be8ba48678e7bf57ec4b385411f"
dependencies = [
 "iana-time-zone",
 "js-sys",
 "num-integer",
 "num-traits",
 "time",
 "wasm-bindgen",
 "winapi",
]

[[package]]
name = "chrono-tz"
version = "0.7.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "bbc529705a6e0028189c83f0a5dd9fb214105116f7e3c0eeab7ff0369766b0d1"
dependencies = [
 "chrono",
 "chrono-tz-build",
 "phf",
]

[[package]]
name = "chrono-tz-build"
version = "0.1.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "d9998fb9f7e9b2111641485bf8beb32f92945f97f92a3d061f744cfef335f751"
dependencies = [
 "parse-zoneinfo",
 "phf",
 "phf_codegen",
]

[[package]]
name = "core-foundation-sys"
version = "0.8.3"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "5827cebf4670468b8772dd191856768aedcb1b0278a04f989f7766351917b9dc"

[[package]]
name = "crc32fast"
version = "1.3.2"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "b540bd8bc810d3885c6ea91e2018302f68baba2129ab3e88f32389ee9370880d"
dependencies = [
 "cfg-if",
]

[[package]]
name = "either"
version = "1.8.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "90e5c1c8368803113bf0c9584fc495a58b86dc8a29edbf8fe877d21d9507e797"

[[package]]
name = "equalto_calc"
version = "0.1.0"
dependencies = [
 "chrono",
 "chrono-tz",
 "once_cell",
 "regex",
 "ryu",
 "serde",
 "serde_json",
 "serde_repr",
]

[[package]]
name = "equalto_xlsx"
version = "0.1.0"
dependencies = [
 "chrono",
 "equalto_calc",
 "itertools",
 "roxmltree",
 "serde",
 "serde_json",
 "thiserror",
 "uuid",
 "zip",
]

[[package]]
name = "flate2"
version = "1.0.24"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "f82b0f4c27ad9f8bfd1f3208d882da2b09c301bc1c828fd3a00d0216d2fbbff6"
dependencies = [
 "crc32fast",
 "miniz_oxide",
]

[[package]]
name = "getrandom"
version = "0.2.8"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "c05aeb6a22b8f62540c194aac980f2115af067bfe15a0734d7277a768d396b31"
dependencies = [
 "cfg-if",
 "libc",
 "wasi 0.11.0+wasi-snapshot-preview1",
]

[[package]]
name = "iana-time-zone"
version = "0.1.48"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "237a0714f28b1ee39ccec0770ccb544eb02c9ef2c82bb096230eefcffa6468b0"
dependencies = [
 "android_system_properties",
 "core-foundation-sys",
 "js-sys",
 "once_cell",
 "wasm-bindgen",
 "winapi",
]

[[package]]
name = "itertools"
version = "0.10.5"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "b0fd2260e829bddf4cb6ea802289de2f86d6a7a690192fbe91b3f46e0f2c8473"
dependencies = [
 "either",
]

[[package]]
name = "itoa"
version = "1.0.3"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "6c8af84674fe1f223a982c933a0ee1086ac4d4052aa0fb8060c12c6ad838e754"

[[package]]
name = "js-sys"
version = "0.3.60"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "49409df3e3bf0856b916e2ceaca09ee28e6871cf7d9ce97a692cacfdb2a25a47"
dependencies = [
 "wasm-bindgen",
]

[[package]]
name = "libc"
version = "0.2.133"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "c0f80d65747a3e43d1596c7c5492d95d5edddaabd45a7fcdb02b95f644164966"

[[package]]
name = "log"
version = "0.4.17"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "abb12e687cfb44aa40f41fc3978ef76448f9b6038cad6aef4259d3c095a2382e"
dependencies = [
 "cfg-if",
]

[[package]]
name = "memchr"
version = "2.5.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "2dffe52ecf27772e601905b7522cb4ef790d2cc203488bbd0e2fe85fcb74566d"

[[package]]
name = "miniz_oxide"
version = "0.5.4"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "96590ba8f175222643a85693f33d26e9c8a015f599c216509b1a6894af675d34"
dependencies = [
 "adler",
]

[[package]]
name = "num-integer"
version = "0.1.45"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "225d3389fb3509a24c93f5c29eb6bde2586b98d9f016636dff58d7c6f7569cd9"
dependencies = [
 "autocfg",
 "num-traits",
]

[[package]]
name = "num-traits"
version = "0.2.15"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "578ede34cf02f8924ab9447f50c28075b4d3e5b269972345e7e0372b38c6cdcd"
dependencies = [
 "autocfg",
]

[[package]]
name = "once_cell"
version = "1.16.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "86f0b0d4bf799edbc74508c1e8bf170ff5f41238e5f8225603ca7caaae2b7860"

[[package]]
name = "parse-zoneinfo"
version = "0.3.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "c705f256449c60da65e11ff6626e0c16a0a0b96aaa348de61376b249bc340f41"
dependencies = [
 "regex",
]

[[package]]
name = "phf"
version = "0.11.1"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "928c6535de93548188ef63bb7c4036bd415cd8f36ad25af44b9789b2ee72a48c"
dependencies = [
 "phf_shared",
]

[[package]]
name = "phf_codegen"
version = "0.11.1"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "a56ac890c5e3ca598bbdeaa99964edb5b0258a583a9eb6ef4e89fc85d9224770"
dependencies = [
 "phf_generator",
 "phf_shared",
]

[[package]]
name = "phf_generator"
version = "0.11.1"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "b1181c94580fa345f50f19d738aaa39c0ed30a600d95cb2d3e23f94266f14fbf"
dependencies = [
 "phf_shared",
 "rand",
]

[[package]]
name = "phf_shared"
version = "0.11.1"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "e1fb5f6f826b772a8d4c0394209441e7d37cbbb967ae9c7e0e8134365c9ee676"
dependencies = [
 "siphasher",
]

[[package]]
name = "proc-macro2"
version = "1.0.43"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "0a2ca2c61bc9f3d74d2886294ab7b9853abd9c1ad903a3ac7815c58989bb7bab"
dependencies = [
 "unicode-ident",
]

[[package]]
name = "quote"
version = "1.0.21"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "bbe448f377a7d6961e30f5955f9b8d106c3f5e449d493ee1b125c1d43c2b5179"
dependencies = [
 "proc-macro2",
]

[[package]]
name = "rand"
version = "0.8.5"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "34af8d1a0e25924bc5b7c43c079c942339d8f0a8b57c39049bef581b46327404"
dependencies = [
 "rand_core",
]

[[package]]
name = "rand_core"
version = "0.6.4"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "ec0be4795e2f6a28069bec0b5ff3e2ac9bafc99e6a9a7dc3547996c5c816922c"

[[package]]
name = "regex"
version = "1.6.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "4c4eb3267174b8c6c2f654116623910a0fef09c4753f8dd83db29c48a0df988b"
dependencies = [
 "aho-corasick",
 "memchr",
 "regex-syntax",
]

[[package]]
name = "regex-syntax"
version = "0.6.27"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "a3f87b73ce11b1619a3c6332f45341e0047173771e8b8b73f87bfeefb7b56244"

[[package]]
name = "roxmltree"
version = "0.13.1"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "dbf7d7b1ea646d380d0e8153158063a6da7efe30ddbf3184042848e3f8a6f671"
dependencies = [
 "xmlparser",
]

[[package]]
name = "ryu"
version = "1.0.11"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "4501abdff3ae82a1c1b477a17252eb69cee9e66eb915c1abaa4f44d873df9f09"

[[package]]
name = "serde"
version = "1.0.144"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "0f747710de3dcd43b88c9168773254e809d8ddbdf9653b84e2554ab219f17860"
dependencies = [
 "serde_derive",
]

[[package]]
name = "serde_derive"
version = "1.0.144"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "94ed3a816fb1d101812f83e789f888322c34e291f894f19590dc310963e87a00"
dependencies = [
 "proc-macro2",
 "quote",
 "syn",
]

[[package]]
name = "serde_json"
version = "1.0.85"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "e55a28e3aaef9d5ce0506d0a14dbba8054ddc7e499ef522dd8b26859ec9d4a44"
dependencies = [
 "itoa",
 "ryu",
 "serde",
]

[[package]]
name = "serde_repr"
version = "0.1.9"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "1fe39d9fbb0ebf5eb2c7cb7e2a47e4f462fad1379f1166b8ae49ad9eae89a7ca"
dependencies = [
 "proc-macro2",
 "quote",
 "syn",
]

[[package]]
name = "siphasher"
version = "0.3.10"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "7bd3e3206899af3f8b12af284fafc038cc1dc2b41d1b89dd17297221c5d225de"

[[package]]
name = "syn"
version = "1.0.100"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "52205623b1b0f064a4e71182c3b18ae902267282930c6d5462c91b859668426e"
dependencies = [
 "proc-macro2",
 "quote",
 "unicode-ident",
]

[[package]]
name = "thiserror"
version = "1.0.35"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "c53f98874615aea268107765aa1ed8f6116782501d18e53d08b471733bea6c85"
dependencies = [
 "thiserror-impl",
]

[[package]]
name = "thiserror-impl"
version = "1.0.35"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "f8b463991b4eab2d801e724172285ec4195c650e8ec79b149e6c2a8e6dd3f783"
dependencies = [
 "proc-macro2",
 "quote",
 "syn",
]

[[package]]
name = "time"
version = "0.1.44"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "6db9e6914ab8b1ae1c260a4ae7a49b6c5611b40328a735b21862567685e73255"
dependencies = [
 "libc",
 "wasi 0.10.0+wasi-snapshot-preview1",
 "winapi",
]

[[package]]
name = "unicode-ident"
version = "1.0.4"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "dcc811dc4066ac62f84f11307873c4850cb653bfa9b1719cee2bd2204a4bc5dd"

[[package]]
name = "uuid"
version = "1.2.2"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "422ee0de9031b5b948b97a8fc04e3aa35230001a722ddd27943e0be31564ce4c"
dependencies = [
 "getrandom",
 "serde",
]

[[package]]
name = "wasi"
version = "0.10.0+wasi-snapshot-preview1"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "1a143597ca7c7793eff794def352d41792a93c481eb1042423ff7ff72ba2c31f"

[[package]]
name = "wasi"
version = "0.11.0+wasi-snapshot-preview1"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "9c8d87e72b64a3b4db28d11ce29237c246188f4f51057d65a7eab63b7987e423"

[[package]]
name = "wasm-bindgen"
version = "0.2.83"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "eaf9f5aceeec8be17c128b2e93e031fb8a4d469bb9c4ae2d7dc1888b26887268"
dependencies = [
 "cfg-if",
 "wasm-bindgen-macro",
]

[[package]]
name = "wasm-bindgen-backend"
version = "0.2.83"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "4c8ffb332579b0557b52d268b91feab8df3615f265d5270fec2a8c95b17c1142"
dependencies = [
 "bumpalo",
 "log",
 "once_cell",
 "proc-macro2",
 "quote",
 "syn",
 "wasm-bindgen-shared",
]

[[package]]
name = "wasm-bindgen-macro"
version = "0.2.83"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "052be0f94026e6cbc75cdefc9bae13fd6052cdcaf532fa6c45e7ae33a1e6c810"
dependencies = [
 "quote",
 "wasm-bindgen-macro-support",
]

[[package]]
name = "wasm-bindgen-macro-support"
version = "0.2.83"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "07bc0c051dc5f23e307b13285f9d75df86bfdf816c5721e573dec1f9b8aa193c"
dependencies = [
 "proc-macro2",
 "quote",
 "syn",
 "wasm-bindgen-backend",
 "wasm-bindgen-shared",
]

[[package]]
name = "wasm-bindgen-shared"
version = "0.2.83"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "1c38c045535d93ec4f0b4defec448e4291638ee608530863b1e2ba115d4fff7f"

[[package]]
name = "winapi"
version = "0.3.9"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "5c839a674fcd7a98952e593242ea400abe93992746761e38641405d28b00f419"
dependencies = [
 "winapi-i686-pc-windows-gnu",
 "winapi-x86_64-pc-windows-gnu",
]

[[package]]
name = "winapi-i686-pc-windows-gnu"
version = "0.4.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "ac3b87c63620426dd9b991e5ce0329eff545bccbbb34f3be09ff6fb6ab51b7b6"

[[package]]
name = "winapi-x86_64-pc-windows-gnu"
version = "0.4.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "712e227841d057c1ee1cd2fb22fa7e5a5461ae8e48fa2ca79ec42cfc1931183f"

[[package]]
name = "xmlparser"
version = "0.13.3"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "114ba2b24d2167ef6d67d7d04c8cc86522b87f490025f39f0303b7db5bf5e3d8"

[[package]]
name = "zip"
version = "0.5.13"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "93ab48844d61251bb3835145c521d88aa4031d7139e8485990f60ca911fa0815"
dependencies = [
 "byteorder",
 "crc32fast",
 "flate2",
 "thiserror",
]
