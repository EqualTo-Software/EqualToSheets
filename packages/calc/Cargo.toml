[workspace]

members = [
  "equalto_calc",
  "equalto_xlsx",
]

exclude = [
  "generate_locale",
]

[profile.release]
lto = true
