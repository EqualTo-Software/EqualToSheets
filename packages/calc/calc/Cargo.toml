[package]
name = "calc"
version = "0.1.0"
authors = ["Nicol√°s Hatcher <nicolas.hatcher@equalto.com>"]
edition = "2021"

# We run `wasm-opt` externally, because the bundled prebuilt binary isn't available for linux/arm64
[package.metadata.wasm-pack.profile.release]
wasm-opt = false

[lib]
crate-type = ["cdylib", "rlib"]

[dependencies]
wasm-bindgen = {version="0.2.63", features = ["serde-serialize"]}
gloo-utils = { version = "0.1.5", features = ["serde"] }
equalto_calc = { path = "../equalto_calc" }
serde = { version = "1.0", features = ["derive"] }
serde_json = "1.0"
js-sys = "0.3.60"


[dev-dependencies]
wasm-bindgen-test = "0.3.13"
