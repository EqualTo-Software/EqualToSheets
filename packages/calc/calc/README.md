CALC
====

## 📚 About

Simple wrapper around `equalto_calc` that compiles to _webassembly_

## 🛠️ Build with `wasm-pack build`

If you want a build for the browser

```
wasm-pack build --target web
```

Will create a `pkg/` folder.

Other deployment options, including how to do it with a bundler like webpack, are discussed here:

[wasm-bidgen deployment](https://rustwasm.github.io/docs/wasm-bindgen/reference/deployment.html)
