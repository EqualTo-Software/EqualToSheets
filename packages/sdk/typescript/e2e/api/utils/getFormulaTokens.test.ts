import { initialize, getApi, FormulaToken, FormulaErrorCode } from '@equalto-software/calc';

describe('getFormulaTokens', () => {
  beforeAll(async () => {
    await initialize();
  });

  test('reference assignment: =D13', async () => {
    const { getFormulaTokens } = (await getApi()).utils;
    expect(getFormulaTokens('=D13')).toEqual<FormulaToken[]>([
      { token: { type: 'COMPARE', data: 'Equal' }, start: 0, end: 1 },
      {
        token: {
          type: 'REFERENCE',
          data: {
            sheet: null,
            row: 13,
            absolute_row: false,
            column: 4,
            absolute_column: false,
          },
        },
        start: 1,
        end: 4,
      },
    ]);
  });

  test('range token: A1:B2', async () => {
    const { getFormulaTokens } = (await getApi()).utils;
    expect(getFormulaTokens('A1:B2')).toEqual<FormulaToken[]>([
      {
        token: {
          type: 'RANGE',
          data: {
            sheet: null,
            left: {
              row: 1,
              absolute_row: false,
              column: 1,
              absolute_column: false,
            },
            right: {
              row: 2,
              absolute_row: false,
              column: 2,
              absolute_column: false,
            },
          },
        },
        start: 0,
        end: 5,
      },
    ]);
  });

  test('references, add/subtract tokens: =A1+B1-$E$14', async () => {
    const { getFormulaTokens } = (await getApi()).utils;
    expect(getFormulaTokens('=A1+B1-$E$14')).toEqual<FormulaToken[]>([
      { token: { type: 'COMPARE', data: 'Equal' }, start: 0, end: 1 },
      {
        token: {
          type: 'REFERENCE',
          data: {
            sheet: null,
            row: 1,
            absolute_row: false,
            column: 1,
            absolute_column: false,
          },
        },
        start: 1,
        end: 3,
      },
      { token: { type: 'SUM', data: 'Add' }, start: 3, end: 4 },
      {
        token: {
          type: 'REFERENCE',
          data: {
            sheet: null,
            row: 1,
            absolute_row: false,
            column: 2,
            absolute_column: false,
          },
        },
        start: 4,
        end: 6,
      },
      { token: { type: 'SUM', data: 'Minus' }, start: 6, end: 7 },
      {
        token: {
          type: 'REFERENCE',
          data: {
            sheet: null,
            row: 14,
            absolute_row: true,
            column: 5,
            absolute_column: true,
          },
        },
        start: 7,
        end: 12,
      },
    ]);
  });

  test('product tokens */: 2*2/2', async () => {
    const { getFormulaTokens } = (await getApi()).utils;
    expect(getFormulaTokens('2*3/4')).toEqual<FormulaToken[]>([
      { token: { type: 'NUMBER', data: 2 }, start: 0, end: 1 },
      { token: { type: 'PRODUCT', data: 'Times' }, start: 1, end: 2 },
      { token: { type: 'NUMBER', data: 3 }, start: 2, end: 3 },
      { token: { type: 'PRODUCT', data: 'Divide' }, start: 3, end: 4 },
      { token: { type: 'NUMBER', data: 4 }, start: 4, end: 5 },
    ]);
  });

  test('misc tokens: ^()[]{}:;,!%&', async () => {
    const { getFormulaTokens } = (await getApi()).utils;
    expect(getFormulaTokens('^()[]{}:;,!%&')).toEqual<FormulaToken[]>([
      { token: { type: 'POWER' }, start: 0, end: 1 },
      { token: { type: 'LPAREN' }, start: 1, end: 2 },
      { token: { type: 'RPAREN' }, start: 2, end: 3 },
      { token: { type: 'LBRACKET' }, start: 3, end: 4 },
      { token: { type: 'RBRACKET' }, start: 4, end: 5 },
      { token: { type: 'LBRACE' }, start: 5, end: 6 },
      { token: { type: 'RBRACE' }, start: 6, end: 7 },
      { token: { type: 'COLON' }, start: 7, end: 8 },
      { token: { type: 'SEMICOLON' }, start: 8, end: 9 },
      { token: { type: 'COMMA' }, start: 9, end: 10 },
      { token: { type: 'BANG' }, start: 10, end: 11 },
      { token: { type: 'PERCENT' }, start: 11, end: 12 },
      { token: { type: 'AND' }, start: 12, end: 13 },
    ]);
  });

  test('identifier tokens: HELLO WORLD', async () => {
    const { getFormulaTokens } = (await getApi()).utils;
    expect(getFormulaTokens('HELLO WORLD')).toEqual<FormulaToken[]>([
      { token: { type: 'IDENT', data: 'HELLO' }, start: 0, end: 5 },
      {
        token: {
          type: 'IDENT',
          data: 'WORLD',
        },
        start: 5, // <- doesn't appear to be correct, shouldn't be 6?
        end: 11,
      },
    ]);
  });

  test('string tokens: "HELLO WORLD"', async () => {
    const { getFormulaTokens } = (await getApi()).utils;
    expect(getFormulaTokens('"HELLO WORLD"')).toEqual<FormulaToken[]>([
      {
        token: {
          type: 'STRING',
          data: 'HELLO WORLD',
        },
        start: 0,
        end: 13,
      },
    ]);
  });

  test.each<[string, number]>([
    ['0', 0],
    ['1', 1],
    ['1234', 1234],
    ['12.34', 12.34],
  ])('number tokens: %s', async (formula, expectedNumber) => {
    const { getFormulaTokens } = (await getApi()).utils;
    expect(getFormulaTokens(formula)).toEqual<FormulaToken[]>([
      { token: { type: 'NUMBER', data: expectedNumber }, start: 0, end: expect.any(Number) },
    ]);
  });

  test.each<[string, number]>([
    ['-0', 0],
    ['-1', 1],
    ['-1234', 1234],
    ['-12.34', 12.34],
  ])('negative numbers return two tokens: %s', async (formula, expectedNumber) => {
    const { getFormulaTokens } = (await getApi()).utils;
    expect(getFormulaTokens(formula)).toEqual<FormulaToken[]>([
      { token: { type: 'SUM', data: 'Minus' }, start: 0, end: 1 },
      { token: { type: 'NUMBER', data: expectedNumber }, start: 1, end: expect.any(Number) },
    ]);
  });

  test('bool tokens: true True TRUE false False FALSE', async () => {
    const { getFormulaTokens } = (await getApi()).utils;
    expect(getFormulaTokens('true True TRUE false False FALSE')).toEqual<FormulaToken[]>([
      { token: { type: 'BOOLEAN', data: true }, start: 0, end: 4 },
      { token: { type: 'BOOLEAN', data: true }, start: 4, end: 9 },
      { token: { type: 'BOOLEAN', data: true }, start: 9, end: 14 },
      { token: { type: 'BOOLEAN', data: false }, start: 14, end: 20 },
      { token: { type: 'BOOLEAN', data: false }, start: 20, end: 26 },
      { token: { type: 'BOOLEAN', data: false }, start: 26, end: 32 },
    ]);
  });

  test('tokens for compare operations', async () => {
    const { getFormulaTokens } = (await getApi()).utils;
    expect(getFormulaTokens('= < <= >= > <>')).toEqual<FormulaToken[]>([
      { token: { type: 'COMPARE', data: 'Equal' }, start: 0, end: 1 },
      { token: { type: 'COMPARE', data: 'LessThan' }, start: 1, end: 3 },
      { token: { type: 'COMPARE', data: 'LessOrEqualThan' }, start: 3, end: 6 },
      { token: { type: 'COMPARE', data: 'GreaterOrEqualThan' }, start: 6, end: 9 },
      { token: { type: 'COMPARE', data: 'GreaterThan' }, start: 9, end: 11 },
      { token: { type: 'COMPARE', data: 'NonEqual' }, start: 11, end: 14 },
    ]);
  });

  test.each<[string, FormulaErrorCode]>([
    ['#REF!', FormulaErrorCode.REF],
    ['#NAME?', FormulaErrorCode.NAME],
    ['#VALUE!', FormulaErrorCode.VALUE],
    ['#DIV/0!', FormulaErrorCode.DIV],
    ['#N/A', FormulaErrorCode.NA],
    ['#NUM!', FormulaErrorCode.NUM],
    ['#ERROR!', FormulaErrorCode.ERROR],
    ['#N/IMPL!', FormulaErrorCode.NIMPL],
    ['#SPILL!', FormulaErrorCode.SPILL],
    ['#CALC!', FormulaErrorCode.CALC],
    ['#CIRC!', FormulaErrorCode.CIRC],
  ])('error tokens: %s', async (formula, errorKind) => {
    const { getFormulaTokens } = (await getApi()).utils;
    expect(getFormulaTokens(formula)).toEqual<FormulaToken[]>([
      { token: { type: 'ERROR', data: errorKind }, start: 0, end: expect.any(Number) },
    ]);
  });

  test('illegal token: emoji', async () => {
    const { getFormulaTokens } = (await getApi()).utils;
    expect(getFormulaTokens('☠️')).toEqual<FormulaToken[]>([
      {
        token: { type: 'ILLEGAL', data: { position: 1, message: 'Unknown error' } },
        start: 0,
        end: 2,
      },
    ]);
  });

  test.each(['=A1+A3', '=$A1+B$2', '=IF($A$1; B4; #N/A)', '=SUM(A3:A4)', "'Sheet name'!A1"])(
    'snapshot test for formula: %s',
    async (formula) => {
      const { getFormulaTokens } = (await getApi()).utils;
      expect(getFormulaTokens(formula)).toMatchSnapshot();
    },
  );
});
