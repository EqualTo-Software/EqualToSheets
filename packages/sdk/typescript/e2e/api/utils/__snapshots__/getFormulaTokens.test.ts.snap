// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getFormulaTokens snapshot test for formula: 'Sheet name'!A1 1`] = `
[
  {
    "end": 15,
    "start": 0,
    "token": {
      "data": {
        "absolute_column": false,
        "absolute_row": false,
        "column": 1,
        "row": 1,
        "sheet": "Sheet name",
      },
      "type": "REFERENCE",
    },
  },
]
`;

exports[`getFormulaTokens snapshot test for formula: =$A1+B$2 1`] = `
[
  {
    "end": 1,
    "start": 0,
    "token": {
      "data": "Equal",
      "type": "COMPARE",
    },
  },
  {
    "end": 4,
    "start": 1,
    "token": {
      "data": {
        "absolute_column": true,
        "absolute_row": false,
        "column": 1,
        "row": 1,
        "sheet": null,
      },
      "type": "REFERENCE",
    },
  },
  {
    "end": 5,
    "start": 4,
    "token": {
      "data": "Add",
      "type": "SUM",
    },
  },
  {
    "end": 8,
    "start": 5,
    "token": {
      "data": {
        "absolute_column": false,
        "absolute_row": true,
        "column": 2,
        "row": 2,
        "sheet": null,
      },
      "type": "REFERENCE",
    },
  },
]
`;

exports[`getFormulaTokens snapshot test for formula: =A1+A3 1`] = `
[
  {
    "end": 1,
    "start": 0,
    "token": {
      "data": "Equal",
      "type": "COMPARE",
    },
  },
  {
    "end": 3,
    "start": 1,
    "token": {
      "data": {
        "absolute_column": false,
        "absolute_row": false,
        "column": 1,
        "row": 1,
        "sheet": null,
      },
      "type": "REFERENCE",
    },
  },
  {
    "end": 4,
    "start": 3,
    "token": {
      "data": "Add",
      "type": "SUM",
    },
  },
  {
    "end": 6,
    "start": 4,
    "token": {
      "data": {
        "absolute_column": false,
        "absolute_row": false,
        "column": 1,
        "row": 3,
        "sheet": null,
      },
      "type": "REFERENCE",
    },
  },
]
`;

exports[`getFormulaTokens snapshot test for formula: =IF($A$1; B4; #N/A) 1`] = `
[
  {
    "end": 1,
    "start": 0,
    "token": {
      "data": "Equal",
      "type": "COMPARE",
    },
  },
  {
    "end": 3,
    "start": 1,
    "token": {
      "data": "IF",
      "type": "IDENT",
    },
  },
  {
    "end": 4,
    "start": 3,
    "token": {
      "type": "LPAREN",
    },
  },
  {
    "end": 8,
    "start": 4,
    "token": {
      "data": {
        "absolute_column": true,
        "absolute_row": true,
        "column": 1,
        "row": 1,
        "sheet": null,
      },
      "type": "REFERENCE",
    },
  },
  {
    "end": 9,
    "start": 8,
    "token": {
      "type": "SEMICOLON",
    },
  },
  {
    "end": 12,
    "start": 9,
    "token": {
      "data": {
        "absolute_column": false,
        "absolute_row": false,
        "column": 2,
        "row": 4,
        "sheet": null,
      },
      "type": "REFERENCE",
    },
  },
  {
    "end": 13,
    "start": 12,
    "token": {
      "type": "SEMICOLON",
    },
  },
  {
    "end": 18,
    "start": 13,
    "token": {
      "data": 4,
      "type": "ERROR",
    },
  },
  {
    "end": 19,
    "start": 18,
    "token": {
      "type": "RPAREN",
    },
  },
]
`;

exports[`getFormulaTokens snapshot test for formula: =SUM(A3:A4) 1`] = `
[
  {
    "end": 1,
    "start": 0,
    "token": {
      "data": "Equal",
      "type": "COMPARE",
    },
  },
  {
    "end": 4,
    "start": 1,
    "token": {
      "data": "SUM",
      "type": "IDENT",
    },
  },
  {
    "end": 5,
    "start": 4,
    "token": {
      "type": "LPAREN",
    },
  },
  {
    "end": 10,
    "start": 5,
    "token": {
      "data": {
        "left": {
          "absolute_column": false,
          "absolute_row": false,
          "column": 1,
          "row": 3,
        },
        "right": {
          "absolute_column": false,
          "absolute_row": false,
          "column": 1,
          "row": 4,
        },
        "sheet": null,
      },
      "type": "RANGE",
    },
  },
  {
    "end": 11,
    "start": 10,
    "token": {
      "type": "RPAREN",
    },
  },
]
`;
